infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/NamespaceNameValidator.java

```diff
@@ -51,6 +53,18 @@ public static boolean isValid(String name) {
     return validateInternal(name).isOk();
   }
 
+  public static String normalize(String namespaceName) {
+    Matcher matcher = NEGATED_NAME_PATTERN.matcher(namespaceName);
+    String replaced = matcher.replaceAll("-");
+    while (!isValid(replaced)) {
```

Comment:
I don't like this potentially infinite loop. Also `isValid` does not work properly, see failing examples above